### ISSUE TITLE
Battle towers unprecise spawn check

### ISSUE THREAD

**Issue by Krampus69** [BrassAmber-Mods/Brass_Amber_BattleTowers#52]
Hi :wave: 
First of all : Thank you so much for this mod !
It is really amazing, thank you thank you ><

All right, there is a problem with spawn check sometimes :
![2022-01-02_01 23 04](https://user-images.githubusercontent.com/81943963/147862999-e815021f-adc4-47eb-8cfc-5c89a54ad2bc.png)

Sometimes the tower will spawn badly, it's actually not that hard to avoid this problem (you can check the ground under the tower + improve the base of the tower with a round ground or "island looking" base for example)

Keep up the great work anyway 👍

**Comment by JMilamber** [comment]
HI! 👋
Thanks for the issue on here and the wonderful picture! :D 
Glad you are enjoying the mod! 😄 

TLDR: Known Issue, Intended to be fixed in 1.18 demo or first 1.18 update, more pictures of failed tower gen = good.

This is a known issue, the code for where a tower spawns is still from the second time I gave it a go.
It currently checks whether 4 block positions are at the same level as the center of the chunk (14 blocks from the center in each direction). This was meant to avoid hills without having to check a large number of blocks (and subsequently slow down world creation) except sometimes the locations it checks are all the same height, even though there is a steep hill there. 

I plan on re-doing the tower generation code for the 1.18 demo release (or at least the first update to it), and more pictures of the generation failing always help! So if you find other unique/wacky spawns of a tower feel free to add a picture of them here.
The other one it is supposed to avoid is water underneath where it spawns.

**Comment by Krampus69** [comment]
![2022-01-03_10 50 48](https://user-images.githubusercontent.com/81943963/147918807-12cb8845-349a-4f3b-89d2-57f0c3417708.png)
![2022-01-03_11 01 44](https://user-images.githubusercontent.com/81943963/147918823-1b4235f0-aa64-4b4c-af0e-f5b293ea41a2.png)
I got some more pictures if you need 👍 Most of the time the spawn is okay but when you couple it with modded biomes with tricky landscape they can become messy
Also, you're talking about the 1.18 version but will the 1.16.5 still be updated when this is fixed ?

**Comment by JMilamber** [comment]
This is awesome! Thanks man!

Is the land tower spawning in modded biomes? I haven't checked since last changing the biome stuff for the land tower.

Hmm, it may be, depends on how much of the code I end up changing to fix this.

**Comment by Krampus69** [comment]
Yes the towers mostly spawn in modded biomes (here it's ByG)
Also, I've seen some towers spawning inside other structures, which is actually pretty cool but I think the NBT has no structure_void block so it cuts right through the structures instead of blending in
![image](https://user-images.githubusercontent.com/81943963/149615088-c2646492-7a48-4859-a2f7-1503958e37d7.png)
![image](https://user-images.githubusercontent.com/81943963/149615094-cbc27ed9-3676-43d8-b914-ad246dc6875f.png)

See there the tower has spawned in a Dungeon Arise structure and it creates a big chunk, would be nice to add structure void outside of the tower (I personnally use structure gel to make it easier maybe you got something else but just in case...)

**Comment by JMilamber** [comment]
Glad their spawning in modded biomes :D 

Yeah, the bottom 2 floors have structure voids around but any floors above that do not. Definitely, something that will be in 1.18 and potentially backported to 1.16.5.
I use worldEdit myself but structure gel looks very enticing, ill have to try it out and see if I like it better that what I am doing with worldedit currently.

**Comment by JMilamber** [comment]
I also had plans to make them not spawn in any other structures by default, with a config option to turn that off and have them spawn anywhere regardless of whether there already is a structure there or not.

**Comment by Asveste** [comment]
![2022-02-13_02 41 58](https://user-images.githubusercontent.com/23436647/153734749-8f0111a2-c9bc-4b17-aa4a-88d73e8d9138.png)

Didn't want to open an issue just for this, because this one right here already exists. And because you said more pictures of failed structure spawning are welcome; there you go!
This one straight up spawned inside one of the Vanilla Mansions.

**Comment by JMilamber** [comment]
Holy cow!  That’s actually really cool!! Haha 😂 I wonder what the chances of this even happening are lol.

**Comment by BlackShadow77** [comment]
Hi!
I know is an old issue but running in a 1.18.1 server (with some friends) i notice a weird generation, nothing in comparison to precedent cases.

![2022-04-21_01 57 53](https://user-images.githubusercontent.com/19372947/165530875-7f3cad19-3ba0-4721-92ff-b92b06e1672c.png)

As you can see beetween mesa and desert biome the land under the structure is dirt! Is this intended or not?

**Comment by JMilamber** [comment]
Ah yes, it both is and isn't. 

TLDR: Grass blocks around tower were intended, spawning in Mesa/desert was an accidental addition. 

For the 1.18.1 build we were originally going to add the desert variant as well, but it got moved to 1.18.2 instead. 
I had still made the biome spawning changes though so towers could spawn in deserts xD. (So not intended) 

However the actual grass blocks directly around the bottom floor are part of the bottom jigsaw piece of the structure. My attempt to have it merge with the ground more favorably. 
(So also kinda intended) 

1.18.2 we removed that extra grass base and are adding a stone brick base that will extend downwards underneath the tower whenever there is extra space.

**Comment by JMilamber** [comment]
![2022-05-01_03 46 13](https://user-images.githubusercontent.com/46755966/166140749-0a581635-d4d8-41b6-af54-8b770509f429.png)

The first step toward fixing this :) 
Eventually, I hope to use noise generation to automatically build/fill in a hill around it when needed. 

For now, though this^ is the fix.
